randovania/game_connection/executor/cs_executor.py

```diff
@@ -0,0 +1,326 @@
+        self.logger = logging.getLogger(type(self).__name__)
+        self._ip = ip
+
+    @property
+    def ip(self) -> str:
+        return self._ip
+
+    @property
+    def lock_identifier(self) -> str | None:
+        return None
+
+    async def connect(self) -> str | None:
+        if self.is_connected():
+            return None
+
+        try:
+            self._socket_error = None
+
+            self.logger.debug("Connecting to %s:%d.", self._ip, self._port)
+            reader, writer = await asyncio.open_connection(self._ip, self._port)
+            self._socket = CSSocketHolder(reader, writer)
+
+            self.logger.debug("Connection open. Requesting API details...")
+            server_info = await self.get_server_info()
+            self.logger.debug(
+                "Server replied with API level %s, platform %s, uuid %s, and offsets %s. Connection successful.",
+                server_info.api_version,
+                server_info.platform,
+                server_info.uuid,
+                server_info.offsets,
+            )
+            self.server_info = server_info
+
+            self.logger.info("Connected")
+
+            return None
+
+        except (
+            OSError,
+            AttributeError,
+            asyncio.TimeoutError,
+            struct.error,
+            UnicodeError,
+            RuntimeError,
+            ValueError,
+            TSCError,
+        ) as e:
+            # UnicodeError is for some invalid ip addresses
+            self._socket = None
+            message = f"Unable to connect to {self._ip}:{self._port} - ({type(e).__name__}) {e}"
+            self._socket_error = e
+            return message
+
+    def disconnect(self) -> None:
+        socket = self._socket
+        self._socket = None
+        if socket is not None:
+            socket.writer.close()
+
+    def is_connected(self) -> bool:
+        return self._socket is not None
+
+    async def _send_request(self, packet: Packet) -> Packet:
+        try:
+            return await self._internal_send_request(packet)
+        except (
+            OSError,
+            AttributeError,
+            asyncio.TimeoutError,
+            struct.error,
+            UnicodeError,
+            RuntimeError,
+            ValueError,
+        ):
+            self.disconnect()
+            raise
+        except TSCError as e:
+            self.logger.debug(e)
+            self.disconnect()
+            raise
+
+    async def _internal_send_request(self, packet: Packet) -> Packet:
+        assert self._socket is not None
+
+        self._socket.writer.write(packet.to_stream)
+        await asyncio.wait_for(self._socket.writer.drain(), timeout=15)
+
+        if packet.type == PacketType.DISCONNECT:
+            return packet  # no response from server when dc'ing
+
+        response = await Packet.from_stream(self._socket.reader)
+        if response.type == PacketType.ERROR:
+            raise TSCError(response.message.decode("cp1252"))
```

Comment:
🙃

Reply:
?

Reply:
Just lamenting the necessity of using bad codecs

Reply:
ah yeah. it was a Windows game in 2004, there's only so much it could do. be glad we don't need to use shift-jis like the *real* vanilla game does